service/storage_proxy.cc

```diff
@@ -2319,6 +2336,31 @@ future<> storage_proxy::send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s,
             allow_hints::no);
 }
 
+future<> storage_proxy::mutate_hint_from_scratch(frozen_mutation_and_schema fm_a_s) {
+    // FIXME: using 1h as infinite timeout. If a node is down, we should get an
+    // unavailable exception.
```

Comment:
A node may go down after we checked for its liveness

Reply:
@vladzcloudius , the timeout was already there, but I'm not sure why it was set to a such high value. Do you remember why it was set like that?

Should we change it to the default write timeout (`write_request_timeout_in_ms`)?

Reply:
I preemptively changed it so that timeout is based on `write_request_timeout_in_ms`.

Reply:
@piodul it was @avikivity who added this timeout value in "storage_proxy: implement write timeouts" patch.

@gleb-cloudius this case is not supposed to be hit when the destination node went DOWN but when it's no longer a replica for the token in question.

Going DOWN is not a justification for sending a hint to all other replicas - it's a totally normal scenario. In such a case we should not attempt sending to the node and wait till it becomes up again: look at `sender::can_send()`.

Reply:
> I preemptively changed it so that timeout is based on `write_request_timeout_in_ms`.

@piodul I'm not sure we want to create a correlation between a user defined write request timeout and hints mutation timeout. Having an infinite timeout here should be totally safe as clarified in Avi's comment in code.

Reply:
No, `mutation` is a one way verb meaning that the sender only waits for message to be sent. Nothing is guarantied about it been processed and reply been eventually received. Sending with infinity timeout here only guaranties that eventually (after running for long enough) we will have infinity hanged writes. Avi lives in a fantasy world where networking is as reliable as shared memory communication.

Reply:
I confirm Gleb's statement. I ran a test where I stopped one node and accumulated hints for it. Then, I started a new node which replaced the node that was down, but quickly stopped the new node while hints were still being sent. I could see that there was a small number of writes that lingered for exactly one hour:

![linger](https://user-images.githubusercontent.com/4950936/73532802-bdf65800-441d-11ea-874d-ff27f20498ab.png)

(I ran scylla without this patch so that it used the 1h timeout)

Reply:
@gleb-cloudius I see. Thanks for a clarification and @piodul thanks for confirming it. 

I wonder why can't we have one of the following:
 * Connect the TCP socket termination and/or marking the remote node as DOWN/"left the ring (there are a few states that denote this)" to reporting a failure on the pending mutations requests sent to that node and waiting for the timeout. The drawback is that after that we may fail much more such requests if networking is flaky but maybe it's not a bad thing. - In this case you prefer to deliver the message about the flakiness to the user as soon as possible.
 * Make this VERB a two-way. Honestly it would make sense to make it such because the semantics of that timeout is kinda imply two-wayness then why not to make it explicit?

Reply:
First thing we may do, but we do not. And yes it will introduce false positives (as in message was handled, but failed it)

Second is because we reuse MUTATE code to send hints and the former is not always point-to-point so it cannot  be two-way. Making hints two-way will require writing a lot of new code for a little benefit. 

And if you cannot deliver a hint in to seconds may be the system has better things to do than to try hard and timeout and retry later is not such a bad idea.

Reply:
Well, then probably we do need it (hints timeout) be configurable. Making it equal to `write_request_timeout_in_ms` is a possibility - probably better than introducing yet another config parameter.